addons/mass_mailing/models/mailing_contact.py

```diff
@@ -31,13 +31,17 @@ class MassMailingContactListRel(models.Model):
     def create(self, vals_list):
         now = fields.Datetime.now()
         for vals in vals_list:
-            if 'opt_out' in vals:
+            if 'opt_out' in vals and 'unsubscription_date' not in vals:
                 vals['unsubscription_date'] = now if vals['opt_out'] else False
+            if 'unsubscription_date' in vals and vals['unsubscription_date']:
+                vals['opt_out'] = True
         return super().create(vals_list)
 
     def write(self, vals):
-        if 'opt_out' in vals:
+        if 'opt_out' in vals and 'unsubscription_date' not in vals:
             vals['unsubscription_date'] = fields.Datetime.now() if vals['opt_out'] else False
+        if 'unsubscription_date' in vals and vals['unsubscription_date']:
```

Comment:
Could be shortened with "if vals.get('unsubscription_date')", as default value of get is None: it checks that value is present and not falsy.